src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,375 @@
+  auto& customEvent = outcome.customEventResult;
+
+  auto serializedResult = rpcResult.getResult().getV8Serialized();
+  auto deserialized = _::deserializeV8(js, kj::heapArray(serializedResult.asBytes()));
+  if (customEvent.outcome == EventOutcome::EXCEPTION) {
+    // The remote Worker threw an exception, so we should throw on the client side too.
+    js.throwException(deserialized);
+  }
+
+  KJ_REQUIRE(customEvent.outcome == EventOutcome::OK, "got unexpected event outcome for WorkerRpc");
+  return jsg::Value(js.v8Isolate, kj::mv(deserialized));
+}
+
 kj::Maybe<jsg::JsValue> WorkerRpc::getNamed(jsg::Lock& js, kj::StringPtr name) {
-    // TODO(soon): The server side PR will fill this function with the RPC call and other stuff,
-    //             this is just a temporary placeholder for testing.
-    return jsg::JsValue(js.wrapReturningFunction(js.v8Context(),
-        [](jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
-          return js.error("WorkerRpc is unimplemented");
+  // Named intercept is enabled, this means we won't default to legacy behavior.
+  // The return value of the function is a promise that resolves once the remote returns the result
+  // of the RPC call.
+  return jsg::JsValue(js.wrapReturningFunction(js.v8Context(), [this, methodName=kj::str(name)]
+      (jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
+        auto& ioContext = IoContext::current();
+        // Wait for the RPC to resolve and then process the result.
+        return js.wrapSimplePromise(ioContext.awaitIo(js, sendWorkerRpc(js, methodName, args),
+            [this] (jsg::Lock& js, WorkerRpc::ResponseAndResult result) -> jsg::Value {
+          return handleRpcResponse(js, kj::mv(result));
+        }));
+      }
+  ));
+}
+
+// The capability that lets us call remote methods over RPC.
+// The client capability is dropped after each callRequest().
+class JsRpcTargetImpl final : public rpc::JsRpcTarget::Server {
+public:
+  JsRpcTargetImpl(
+      kj::Own<kj::PromiseFulfiller<bool>> callFulfiller,
+      IoContext& ctx,
+      kj::Maybe<kj::StringPtr> entrypointName)
+      : callFulfiller(kj::mv(callFulfiller)), ctx(ctx), entrypointName(entrypointName) {}
+
+  // Handles the delivery of JS RPC method calls.
+  kj::Promise<void> call(CallContext callContext) override {
+    auto methodName = kj::heapString(callContext.getParams().getMethodName());
+    auto serializedArgs = callContext.getParams().getSerializedArgs().getV8Serialized().asBytes();
+
+    bool caughtException = false;
+    // Try to execute the requested method.
+    try {
+      caughtException = co_await ctx.run(
+          [this, methodName=kj::mv(methodName), serializedArgs = kj::mv(serializedArgs),
+           entrypointName = entrypointName, &callContext] (Worker::Lock& lock) ->
+              kj::Promise<bool> {
+
+        bool exceptionThrown = false;
+        auto reader = callContext.initResults(capnp::MessageSize { 4, 1 }).initResult();
```

Comment:
`MessageSize { 4, 1 }` does not seem like an appropriate size estimate here. The size depends on the serialized V8 size. I'd suggest serializing to a byte array first, then using that to compute the size estimate here. Remember that the first number is a size in _words_, where each word is 8 bytes, so divide the byte size by 8. I'd also add 8 or so words for struct encoding overhead. For the second number, we're not returning any capabilities, so it can be zero.

```suggestion
        auto reader = callContext.initResults(capnp::MessageSize { serializedBytes.size() / 8 + 8, 0 }).initResult();
```

Reply:
Gave this a go, it works fine until we're handling a promise on the server side. After we `co_await` the call to `prepareAsyncResult()` [here](https://github.com/cloudflare/workerd/pull/1028/files#diff-bbe685e15c44d4de0aed0b4b484f1c0fbe1427fccbf80ab9f65c6180f6632203R154), the `CallContext` seems to be destroyed and so further access attempts segfault. Since we don't know the size of the result until the promise resolves we have to wait for the promise first, but by then it's too late.

It looks like allocating a builder _before_ we go async on that line prevents `CallContext` from being dtor'd, so I'm wondering if it's alright to just allocate the `Builder` without a `sizeHint` before we invoke the JS function? What is the consequence of using the default `sizeHint`?

Reply:
Something else must be wrong here -- the `CallContext` is definitely not destroyed until the RPC call completes. What exactly did you see?

It's wasteful to allocate the results object before you're ready to initialize it. Best practice is to not call `initResults()` until the RPC is just about done and ready to write the results.

If you allocate without a size hint, the default behavior will probably allocate a lot more space than you need, which is also wasteful. (That said, I would like to come up with an adaptive algorithm that avoids this problem... someday...)

Reply:
Removed some code for brevity but:

```
kj::Promise<void> call(CallContext callContext) override {
    // Try to execute the requested method.
    try {
        // Not everything is captured here to keep ex. shorter.
        co_await ctx.run([this, &callContext] (Worker::Lock& lock) -> kj::Promise<bool> {
            jsg::Lock& js = lock;

            kj::Maybe<SerializedResult> eventualResult;
            if (fn->IsAsyncFunction()) {
                // If we enter this and go async, attempting to 'initResults()'
                // the callContext will segfault.
                eventualResult = co_await prepareAsyncResult(js, fn, serializedArgs);
            } else {
                // If we enter this then attempting to `initResults()`
                // on callContext will succeed.
                eventualResult = prepareSyncResult(js, fn, serializedArgs);
            }
            auto result = kj::mv(KJ_REQUIRE_NONNULL(eventualResult));

            // NEXT LINE TRIGGERS SEGFAULT.
            auto builder = callContext.initResults();
            auto final = builder.initResult();
            final.setV8Serialized(kj::mv(result.data));
            co_return exceptionThrown;
        });
    } catch(kj::Exception e) {
        if (auto desc = e.getDescription();
            !jsg::isTunneledException(desc) && !jsg::isDoNotLogException(desc)) {
            LOG_EXCEPTION("JsRpcTargetCall"_kj, e);
        }
    }
    co_return;
}
```

I can push a fixup here to move the `initResults()` call down as you mentioned in the initial comment in this thread, and then also turn the `actor-stub` test's method `async`, it should segfault in CI.

Also, part of the stack trace:
```
----------------------------------- stack #1 out of 1 (22c2bdc8) ---------------------------------
    ??:0:0 ? ?
    
    /proc/self/cwd/bazel-out/k8-dbg/bin/external/capnp-cpp/src/capnp/_virtual_includes/capnp/capnp/capability.h:1153:27 capnp::CallContext<workerd::rpc::JsRpcTarget::CallParams,
                       workerd::rpc::JsRpcTarget::CallResults>::
        initResults(kj::Maybe<capnp::MessageSize>)
    
    /proc/self/cwd/external/workerd/src/workerd/api/worker-rpc.c++:173:36 workerd::api::JsRpcTargetImpl::call(
        capnp::CallContext<workerd::rpc::JsRpcTarget::CallParams,
                           workerd::rpc::JsRpcTarget::CallResults>)::
        'lambda'(workerd::Worker::Lock &)::operator()(
            workerd::Worker::Lock &)(.resume)
```

(following coroutine resume).

Reply:
Alright, every build passed except `windows-debug`, which failed on `js-rpc-disabled.wd-test` [mentioned here](https://github.com/cloudflare/workerd/pull/1028#discussion_r1394711096). I'll be sure to look into that, but I want to push the fixup I mentioned above.

Reply:
Your problem in that snippet is that you're capturing `callContext` by reference, but it is a local variable, so that reference is invalid by the time the lambda runs. You can capture it by copy instead, i.e. remove the `&` prefix in the capture declaration.

Reply:
(To clarify, `CallContext` has pointer-like semantics. So it can be copied, but it doesn't actually own the underlying context.)

Reply:
Still segfaults in the same spot, I had tried `kj::mv`ing it before (and just tried again) too, but that also results in the same segfault :/

It's weird that linux asan passes, while debug and the regular test fail. I'll push the capture by copy to see if that changes anything in CI but locally I still see sigsegv. Will keep looking while that runs.

Reply:
The other problem is that this is a lambda coroutine with captures. You need to wrap the lambda in `kj::coCapture()`, otherwise all the captures become invalid after the first `co_await`.

Reply:
Thank you, this fixed it!

I suddenly have a reason to re-read the kj tour doc, specifically [coroutines](https://github.com/capnproto/capnproto/blob/master/kjdoc/tour.md#coroutines) :)